main/src/ca/uwaterloo/flix/util/collection/Chain.scala

```diff
@@ -43,10 +43,81 @@ sealed trait Chain[+A] extends Iterable[A] {
     }
   }
 
+  def isEmpty: Boolean = this match {
+    case Chain.Empty => true
+    case Chain.Link(l, r) => l.isEmpty && r.isEmpty
+    case Chain.Many(cs) => cs.forall(_.isEmpty)
+    case Chain.Proxy(xs) => xs.isEmpty
+  }
+
+  def head: Option[A] = this match {
+    case Chain.Empty => None
+    case Chain.Link(l, _) => l.head
+    case Chain.Many(cs) => cs.find(_.head.isDefined).flatMap(_.head)
+    case Chain.Proxy(xs) => xs.headOption
+  }
+
+  /**
+    * Returns `this` as a [[List]].
+    */
+  def toList: List[A] = this match {
+    case Chain.Empty => List.empty
+    case Chain.Link(l, r) => l.toList ++ r.toList
+    case Chain.Many(cs) => cs.flatMap(_.toList).toList
```

Comment:
I think this flatMap call is sus. 

But even if it is not, its not great performance. 

I suggest, lets create a ListBuffer, call foreach on this and add each element to the buffer. Then turn the buffer into a list.